db/write_thread.cc

```diff
@@ -558,18 +558,42 @@ void WriteThread::ExitAsBatchGroupLeader(WriteGroup& write_group,
     if (!leader->ShouldWriteToMemtable()) {
       CompleteLeader(write_group);
     }
+
+    // We have to link our group to memtable writer queue before wake up the
+    // next leader or set newest_writer_ to null, otherwise the next leader
+    // can run ahead of us and link to memtable writer queue before we do.
+    //
+    // However, last_writer can finish writing to memtable and join
+    // newest_writer_ queue again before we look for next leader. As a result,
+    // when we scan the queue and look for next leader, if we encounter
+    // last_writer, we won't be able to tell if we reach the current write
+    // group or not. Push a dummy writer to the queue so we can tell the two
+    // cases apart.
+    Writer dummy;
+    Writer* group_boundary = last_writer;
+    Writer* newest_writer = last_writer;
+    if (newest_writer_.compare_exchange_strong(newest_writer, &dummy)) {
+      group_boundary = &dummy;
+      newest_writer = &dummy;
+    }
+
     // Link the ramaining of the group to memtable writer list.
     if (write_group.size > 0) {
       if (LinkGroup(write_group, &newest_memtable_writer_)) {
         // The leader can now be different from current writer.
         SetState(write_group.leader, STATE_MEMTABLE_WRITER_LEADER);
       }
     }
-    // Reset newest_writer_ and wake up the next leader.
-    Writer* newest_writer = last_writer;
-    if (!newest_writer_.compare_exchange_strong(newest_writer, nullptr)) {
+
+    // If group_boundary is last_writer, start to look for next leader from
+    // newest_writer. Otherwise, try to reset newest_writer_ to null.
+    bool has_pending_writer =
+        group_boundary != &dummy ||
+        !newest_writer_.compare_exchange_strong(newest_writer, nullptr);
```

Comment:
1. I guess "newest_writer_.compare_exchange_strong(last_writer, nullptr)" might be easier to read
2. I do not see the use of newest_writer variable here. If we get to this compare_exchange_strong, newest_writer is equal to last_writer.
3. If above is correct, then we call the 2nd compare_exchange_strong only if the first one has failed before, and if the first one is failed already, meaning that newest_writer_ is no longer equal to last_writer, how could the 2nd compare_exchange_strong would work: if it was not equal the first time, it would not be equal in the 2nd time either!

Reply:
To make it easier to read, can you replace it with:
```
bool has_pending_writer = true; // When you define dummy
has_pending_writer = false; // When replacing with dummy works above;
if (!has_pending_writer) {
  has_pending_writer = newest_writer_.compare_exchange_strong(newest_writer, nullptr);
}
```

Reply:
Let me explain a bit more. There could be two cases on the first compare_exchange_strong:
1. newest_writer_ == last_writer, in this case there isn't pending writers at that point.
    we append dummy after last_writer. newest_writer does not mean anything, but I set it to dummy for the second compare_exchange_strong.
2. newest_writer_ != last_writer, in this case there's at least one pending writer.
     In this case newest_writer is set to be newest_writer_ (sorry for the confusing naming).

We need the second compare_exchange_strong only in case 1, where we are not sure if there's pending writer enter when we call LinkGroup. The second compare_exchange_strong is used to double check. In both cases, at the end newest_writer (no underscore) points to one of the pending writer, and we search for next leader from there.

Reply:
Regarding your comments:
2. newest_writer is used below at line 595. We search from newest_writer following link_older to find next leader.
3. actually we call the 2nd compare_exchange_strong only if the first one succeeded (case 1 above). In case 2 the first compare_exchange_strong already return a good pending writer and store in newest_writer.